cranelift-wasm/src/sections_translator.rs

```diff
@@ -338,3 +339,65 @@ pub fn parse_data_section<'data>(
 
     Ok(())
 }
+
+/// Parses the Name section of the wasm module.
+pub fn parse_name_section<'data>(
+    names: NameSectionReader<'data>,
+    environ: &mut dyn ModuleEnvironment<'data>,
+) -> WasmResult<()> {
+    for subsection in names {
+        match subsection? {
+            wasmparser::Name::Module(module_name) => {
+                let module_name = module_name.get_name()?;
+                environ.declare_module_name(module_name);
+            }
+            wasmparser::Name::Function(function_names) => {
+                let mut naming_reader = function_names.get_map()?;
+                let mut previous_indices: HashSet<u32> = HashSet::new();
+                let mut previous_names: HashSet<&str> = HashSet::new();
+                for _ in 0..naming_reader.get_count() {
+                    let Naming { index, name } = naming_reader.read()?;
+                    if previous_indices.insert(index) && previous_names.insert(name) {
```

Comment:
Here's where the validation previously discussed happens! After initialising two empty hash sets for names/indices, we check that each new `Naming` does not include a name/index that has been seen before. (i.e., `insert` returns true)

Reply:
Tangentially, I hoped to initialise these `previous_*` sets via `HashSet::with_capacity` but `reader.get_count` returns a `usize`. Converting would mean using `try_into().unwrap()`, and introducing the potential for error didn't seem worth the small perf gain.